rs/registry/canister/canister/canister.rs

```diff
@@ -186,34 +185,63 @@ ic_nervous_system_common_build_metadata::define_get_build_metadata_candid_method
 
 #[export_name = "canister_query get_changes_since"]
 fn get_changes_since() {
-    let response_pb = match deserialize_get_changes_since_request(arg_data()) {
-        Ok(version) => {
-            let registry = registry();
+    fn main() -> Result<RegistryGetChangesSinceResponse, (Code, String)> {
+        // Only self-authenticating and anonymous principals are allowed to call us.
+        const ALLOWED_CALLER_CLASSES: [Result<PrincipalIdClass, String>; 2] = [
+            Ok(PrincipalIdClass::SelfAuthenticating),
+            Ok(PrincipalIdClass::Anonymous),
+        ];
+        let class = caller().class();
+        if !ALLOWED_CALLER_CLASSES.contains(&class) {
```

Comment:
were you going to use try_from_principal_id here somewhere?

Reply:
Yes, I did, but Martin asked me to change this.

Reply:
I suggested to explicitly specify allowed classes instead of disallowed ones (this way, any potential new classes are disallowed by default which is more secure). Since the function `try_from_principal_id` checks a few more conditions than captured by the classes, we could additionally call it for principals of the canister ID class.